Mainnet/HashBattle/HashBattle/Arena.cs

```diff
@@ -0,0 +1,194 @@
+    {
+        Assert(Message.Sender == BattleOwner, "Only battle owner can end game.");
+
+        var battle = GetBattle(battleId);
+
+        Assert(battle.Winner == Address.Zero, "Battle ended.");
+
+        var user = GetUser(battleId, userAddress);
+
+        Assert(!user.ScoreSubmitted, "The user already submitted score.");
+
+        user.Score = score;
+        user.ScoreSubmitted = true;
+
+        SetUser(battleId, userAddress, user);
+
+        if (IsBattleOver)
+            ProcessWinner(battle);
+
+        Log(user);
+        return true;
+    }
+
+    /// <summary>
+    /// Get winner address
+    /// </summary>
+    public Address GetWinner(ulong battleId)
+    {
+        var battle = GetBattle(battleId);
+        Log(battle);
+        return battle.Winner;
+    }
+
+    /// <summary>
+    /// Process winner when all user scores are submitted
+    /// </summary>
+    private void ProcessWinner(BattleMain battle)
+    {
+        if (battle.Users.Length <= 4)
+        {
+            foreach (Address userAddress in battle.Users)
+            {
+                var user = GetUser(battle.BattleId, userAddress);
+                if (!user.ScoreSubmitted)
+                    return;
+            }
+        }
+        uint winnerIndex = GetWinnerIndex(battle.BattleId, battle.Users);
+        if (battle.Winner == Address.Zero)
+        {
+            battle.Winner = battle.Users[winnerIndex];
+            SetBattle(battle.BattleId, battle);
+            ProcessPrize(battle.BattleId);
+        }
+    }
+
+    /// <summary>
+    /// Get winner user index from battle users
+    /// </summary>
+    private uint GetWinnerIndex(ulong battleid, Address[] users)
+    {
+        uint winningScore = 0;
+        uint winningScoreIndex = 0;
+        for (uint i = 0; i < users.Length; i++)
+        {
+            var user = GetUser(battleid, users[i]);
+            if (user.Score > winningScore)
+            {
+                winningScore = user.Score;
+                winningScoreIndex = i;
+            }
+        }
+        return winningScoreIndex;
+    }
+
+    /// <summary>
+    /// Send 3/4 amount to winner and 1/4 amount to battle owner
+    /// </summary>
+    private void ProcessPrize(ulong battleid)
+    {
+        var battle = GetBattle(battleid);
+        ulong prize = battle.Fee * (battle.MaxUsers - 1);
+        Transfer(battle.Winner, prize);
+        Transfer(BattleOwner, battle.Fee);
+    }
+
+    private void SetUser(ulong battleid, Address address, BattleUser user)
+    {
+        PersistentState.SetStruct($"user:{battleid}:{address}", user);
+    }
+
+    private BattleUser GetUser(ulong battleid, Address address)
+    {
+        return PersistentState.GetStruct<BattleUser>($"user:{battleid}:{address}");
+    }
+
+    private void SetBattle(ulong battleid, BattleMain battle)
+    {
+        PersistentState.SetStruct($"battle:{battleid}", battle);
+    }
+
+    private BattleMain GetBattle(ulong battleid)
```

Comment:
Consider making this public so anyone can determine the state of an ongoing battle without having to analyze and parse logs